py3status/py3.py

```diff
@@ -1008,3 +1008,73 @@ def request(self, url, params=None, data=None, headers=None,
                             headers=headers,
                             timeout=timeout,
                             auth=auth)
+
+    def _val_to_bar(self, value, min_val=0, max_val=100):
+        """
+        Create a bar of the appropriate height for each value and adjusts the
+        colors depending on the thresholds.
+        There are 8 possible levels.
+        By default, this assumes values are percentages, but the min/max can be
+        adjusted.
+
+        :param value: value to transform to a bar
+        :param min_val: minimum value possible
+        :param max_val: maximum value possible
+
+        :returns: Composite
+        """
+
+        bars = ['▁', '▂', '▃', '▄', '▅', '▆', '▇', '█']
+
+        idx = int((value - min_val) / (max_val - min_val) * len(bars))
+        if idx == len(bars):
+            idx -= 1
+        full_text = bars[idx]
+        color = self.threshold_get_color(value)
+        return {'full_text': full_text, 'color': color}
+
+    def concurrent_bar_graphs(self, values, min_val=0, max_val=100):
+        """
+        Transforms a list of value into a list of bar Composites.
+
+        :param values: list of values to transform into a bar graph
+        :param min_val: minimum value possible
+        :param max_val: maximum value possible
+
+        :returns: Composite list
+        """
+
+        val_bars = []
+        for val in values:
+            val_bars.append(self._val_to_bar(val,
+                                             min_val=min_val,
+                                             max_val=max_val))
+
+        return self.composite_create(val_bars)
+
+    def history_bar_graph(self, history, value, length=5, default_value=0, min_val=0, max_val=100):
+        """
+        Transform a value into a bar graph with history.
+
+        :param history: history of the variable
```

Comment:
I tried both, but figured it would be better to have the module handle the history. That way, they can tinker with it, and if the module has several such graphs, we're sure there won't be any collision.

I'm also unsure whether we want to store the bar or the raw value. If we store the bar, that's less computation, but if we store the value, then we could change the min/max halfway through and recompute the right bar (this could be used for networks, where there isn't really a maximum, so we could make it dynamic depending on the last few values or something). It also makes it easier for the module to tinker with the history if they have access to the raw values.

Reply:
>I'm also unsure whether we want to store the bar or the raw value...

Could be decided based on `history`. If `history=None`, use bar. Otherwise, use raw.

Reply:
if `history=None` where do we store it? I think storing the raw value is fine, creating the composites each time shouldn't be too taxing.

Reply:
In same place you put `self._default_bars`. Call it `self._index_cache, I guess.

I think in the long run, we might want to keep `bars` inside the modules. (Only 1 so far)... so maybe no more `self._default_bars`... kinda like `clock.py` and its blocks.